### DISCUSSION TITLE
query/mutation cache 'onSuccess', 'onError', etc handlers cannot access the queryClient.

### DISCUSSION THREAD

**Discussion by lm-chad-simmons** [TanStack/query#3450] [Ideas]
I'm building a simplified model CRUD framework using react query, and I have a need to configure a number of defaults for various queries and mutations. Specifically I'd like to have the mutationCache invalidate queries in the onSuccess handler (to update query results that are filtered based on data that has been mutated).  Since the mutationCache and queryCache both need to be created (and configured with the onSuccess handler) before dropping them into the queryClient configuration I don't have access to the queryClient when I'm configuring the mutation cache.  

I have thought of a couple of options for working around the issue:
- use a mutable container to hold the query client and have the onSuccess handler indirectly access the client via this container (this feels kludgy)
- specify the onSuccess handler in the queryClient via setMutationDefaults (this doesn't work if the use overrides the onSuccess handler for a context specific handler for notification messaging etc)

The latter makes for a leaky abstraction, while the former just seems like a hack. The real fix for this would be to provide the queryClient to the onSuccess (and other) handlers so they can properly update the cache's as needed.

**Comment by TkDodo** [comment]
This seems to work fine:

```
const queryClient = new QueryClient();

queryClient.getQueryCache().config.onSuccess = (query) => {
  console.log(query);
};
```

TypeScript is also fine with it because config is a public property.

**Reply by lm-chad-simmons** [reply]
I'm fine with that answer, I just didn't want to rely too heavily on the internal organization of the query client or query cache. If the implementation changes it might not work in a later version.

**Reply by TkDodo** [reply]
true, but public properties are not consistent internal. If we change those, it's a breaking change as if we'd change a top level queryClient function
